diff -uNr glibc-2.20/sysdeps/generic/ldsodefs.h glibc-2.20+patched/sysdeps/generic/ldsodefs.h
--- glibc-2.20/sysdeps/generic/ldsodefs.h	2014-09-07 09:09:09.000000000 +0100
+++ glibc-2.20+patched/sysdeps/generic/ldsodefs.h	2014-12-13 11:44:23.265482088 +0000
@@ -1035,6 +1035,12 @@
 /* Initialization which is normally done by the dynamic linker.  */
 extern void _dl_non_dynamic_init (void) internal_function;
 
+/* We need special support to initialize DSO loaded for statically linked
+   binaries.  */
+extern void _dl_static_init (struct link_map *map);
+#undef DL_STATIC_INIT
+#define DL_STATIC_INIT(map) _dl_static_init (map)
+
 /* Used by static binaries to check the auxiliary vector.  */
 extern void _dl_aux_init (ElfW(auxv_t) *av) internal_function;
 
diff -uNr glibc-2.20/sysdeps/unix/sysv/linux/dl-static.c glibc-2.20+patched/sysdeps/unix/sysv/linux/dl-static.c
--- glibc-2.20/sysdeps/unix/sysv/linux/dl-static.c	1970-01-01 01:00:00.000000000 +0100
+++ glibc-2.20+patched/sysdeps/unix/sysv/linux/dl-static.c	2014-11-28 17:01:38.834800216 +0000
@@ -0,0 +1,69 @@
+/* Variable initialization.
+   Copyright (C) 2001, 2002, 2003, 2004, 2010 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, write to the Free
+   Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
+   02111-1307 USA.  */
+
+#include <ldsodefs.h>
+
+#ifdef SHARED
+
+void
+_dl_var_init (void *array[])
+{
+  /* It has to match "variables" below. */
+  enum
+    {
+      DL_PAGESIZE = 0,
+      DL_CLKTCK
+    };
+
+ GLRO(dl_pagesize) = *((size_t *) array[DL_PAGESIZE]);
+  GLRO(dl_clktck) = *((int *) array[DL_CLKTCK]);
+}
+
+#else
+#include <bits/libc-lock.h>
+
+__libc_lock_define_initialized_recursive (static, _dl_static_lock)
+
+static void *variables[] =
+{
+  &GLRO(dl_pagesize),
+  &GLRO(dl_clktck)
+};
+
+void
+_dl_static_init (struct link_map *map)
+{
+  const ElfW(Sym) *ref = NULL;
+  lookup_t loadbase;
+  void (*f) (void *[]);
+
+  __libc_lock_lock_recursive (_dl_static_lock);
+
+  loadbase = _dl_lookup_symbol_x ("_dl_var_init", map, &ref,
+				  map->l_local_scope, NULL, 0, 1, NULL);
+  if (ref != NULL)
+    {
+      f = (void (*) (void *[])) DL_SYMBOL_ADDRESS (loadbase, ref);
+      f (variables);
+    }
+
+  __libc_lock_unlock_recursive (_dl_static_lock);
+}
+
+#endif
diff -uNr glibc-2.20/sysdeps/unix/sysv/linux/Makefile glibc-2.20+patched/sysdeps/unix/sysv/linux/Makefile
--- glibc-2.20/sysdeps/unix/sysv/linux/Makefile	2014-09-07 09:09:09.000000000 +0100
+++ glibc-2.20+patched/sysdeps/unix/sysv/linux/Makefile	2014-11-28 17:13:49.034795764 +0000
@@ -176,7 +176,9 @@
 
 ifeq ($(subdir),elf)
 sysdep-rtld-routines += dl-brk dl-sbrk dl-getcwd dl-openat64 dl-opendir \
-			dl-fxstatat64
+			dl-fxstatat64 dl-static
+sysdep-dl-routines += dl-static
+sysdep_routines += dl-static
 
 CPPFLAGS-lddlibc4 += -DNOT_IN_libc
 
